### PR TITLE
Support disconnect/reconnect with Quorum DDS

### PR DESCRIPTION
**Pull request by ChumpChief**
[AB#327](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/327)

This change adds support for disconnect/reconnect to the Quorum DDS.  Because the resolution strategy of the Quorum already waits to hear back from the server and acknowledges that local attempts to propose may fail, this is already mostly supported.  The main change in this PR is to be selective about resubmission of ops on reconnect to simplify tracking of accepts and pending values.

Further test mock improvements tracked in [AB#535](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/535)

~~Unfortunately our testing infrastructure doesn't appear to support testing of disconnect/reconnect flows particularly well outside of last-write-wins DDSs.  I add a bit of functionality to the test utils to enable more fine-grained control of disconnect timing relative to ops, and also more-realistic QuorumClients behavior so I can test a few scenarios.  The rest of the intended test scenarios I've stubbed out, but will require a more-substantial investment in our testing infrastructure to implement.~~

### CODE REVIEW

File: packages/runtime/test-runtime-utils/src/mocksForReconnection.ts [hunk 1]
```diff
@@ -33,15 +33,17 @@ export class MockContainerRuntimeForReconnection extends MockContainerRuntime {
             this.clientId = uuid();
             // Update the clientId in FluidDataStoreRuntime.
             this.dataStoreRuntime.clientId = this.clientId;
+            this.factory.quorum.addMember(this.clientId, {});
```
Comment (ChumpChief) [inline packages/runtime/test-runtime-utils/src/mocksForReconnection.ts:36]: This is still probably not sufficient -- with a shared quorum, offline clients will still see live updates to the quorum.

Comment (ChumpChief) [reply packages/runtime/test-runtime-utils/src/mocksForReconnection.ts:45]: Not planning to improve in this PR, tracking this improvement in AB#535

**Review by Josmithr — CHANGES_REQUESTED** [review]

File: packages/dds/quorum/src/test/quorum.spec.ts [hunk 2]
```diff
@@ -325,22 +325,103 @@ describe("Quorum", () => {
...
+            containerRuntime2.connected = true;
+            // Processing an unexpected accept will error and fail the test
+            containerRuntimeFactory.processAllMessages();
         });
 
-        describe("Behavior while disconnected", () => {
-            it("Can do something", async () => {
-                assert.strict(true);
-            });
+        // eslint-disable-next-line max-len
```
[the remark this answers is not in the document]
Comment (ChumpChief) [reply packages/dds/quorum/src/test/quorum.spec.ts:342]: I like the max-len rule generally, even in tests.  I only like overriding it where I think breaking the line would be worse for readability than keeping it long (which is mostly just long strings IMO).  So I'm generally fine with single-line disables.

File: packages/dds/quorum/src/test/quorum.spec.ts [hunk 3]
```diff
@@ -325,22 +325,103 @@ describe("Quorum", () => {
             quorum2.connect(services2);
         });
 
-        describe("Behavior transitioning to disconnect", () => {
-            it("Can do something", async () => {
-                assert.strict(true);
-            });
+        // TODO: Consider if there's any value in distinctly testing these scenarios for acceptance via
```
[the remark this answers is not in the document]
Comment (ChumpChief) [reply packages/dds/quorum/src/test/quorum.spec.ts:328]: Done in latest

File: packages/runtime/test-runtime-utils/src/mocks.ts [hunk 4]
```diff
@@ -230,19 +230,37 @@ export class MockContainerRuntimeFactory {
         this.messages.push(msg as ISequencedDocumentMessage);
     }
 
-    public processAllMessages() {
-        while (this.messages.length > 0) {
-            let msg = this.messages.shift();
+    public processOneMessage() {
```
[the remark this answers is not in the document]
Comment (ChumpChief) [reply packages/runtime/test-runtime-utils/src/mocks.ts:242]: Added in latest

**Review by Josmithr — APPROVED** [review]

**Pull request merged**